odl/discr/partition.py

```diff
@@ -766,6 +767,132 @@ def uniform_partition_fromgrid(grid, begin=None, end=None):
+
+    Notes
+    -----
+    In each axis, 3 of the 4 possible parameters must be given. If all
+    four are provided, they are checked for consistency.
+
+    See also
+    --------
+    uniform_partition :
+        equivalent if ``begin``, ``end`` and ``num_nodes`` are given
+    uniform_partition_fromintv : partition an existing set
+    uniform_partition_fromgrid : use an existing grid as basis
+
+    Examples
+    --------
+    Any combination of three of the four parameters can be used for
+    creation of a partition:
+
+    >>> part = uniform_partition_options(begin=0, end=1, num_nodes=4)
+    >>> part.cell_boundary_vecs
+    (array([ 0.  ,  0.25,  0.5 ,  0.75,  1.  ]),)
+    >>> part = uniform_partition_options(begin=0, num_nodes=4, cell_sides=0.5)
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ,  1.5,  2. ]),)
+    >>> part = uniform_partition_options(end=1, num_nodes=4, cell_sides=0.5)
+    >>> part.cell_boundary_vecs
+    (array([-1. , -0.5,  0. ,  0.5,  1. ]),)
+    >>> part = uniform_partition_options(begin=0, end=1, cell_sides=0.5)
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ]),)
+
+    In higher dimensions, the parameters can be given differently in
+    each axis. Where None is given, the value will be computed:
+
+    >>> part = uniform_partition_options(begin=[0, 0], end=[1, 2],
+    ...                                  num_nodes=[4, 2])
+    >>> part.cell_boundary_vecs
+    (array([ 0.  ,  0.25,  0.5 ,  0.75,  1.  ]), array([ 0.,  1.,  2.]))
+    >>> part = uniform_partition_options(begin=[0, 0], end=[1, 2],
+    ...                                  num_nodes=[None, 2],
+    ...                                  cell_sides=[0.5, None])
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ]), array([ 0.,  1.,  2.]))
+    >>> part = uniform_partition_options(begin=[0, None], end=[None, 2],
+    ...                                  num_nodes=[2, 2],
+    ...                                  cell_sides=[0.5, 0.25])
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ]), array([ 1.5 ,  1.75,  2.  ]))
+    """
+    # np.size(None) = 1
+    sizes = [np.size(p) for p in (begin, end, num_nodes, cell_sides)]
+    max_size = int(np.max(sizes))
+
+    if max_size == 1:
+        begin = [begin]
+        end = [end]
+        num_nodes = [num_nodes]
+        cell_sides = [cell_sides]
+    else:
+        if begin is None:
+            begin = [None] * max_size
+        if end is None:
+            end = [None] * max_size
+        if num_nodes is None:
+            num_nodes = [None] * max_size
+        if cell_sides is None:
+            cell_sides = [None] * max_size
+
+        begin = list(begin)
+        end = list(end)
+        num_nodes = list(num_nodes)
+        cell_sides = list(cell_sides)
+
+    for i, (b, e, n, s) in enumerate(zip(begin, end, num_nodes, cell_sides)):
+        num_params = sum(p is not None for p in (b, e, n, s))
+        if num_params < 3:
```

Comment:
Do we really want to accept anything else than 3? 4 parameters will only lead to confusion, no?